airflow/models/baseoperator.py

```diff
@@ -60,9 +60,25 @@
 ScheduleInterval = Union[str, timedelta, relativedelta]
 
 
+class BaseOperatorMeta(type):
+    """
+    Base metaclass of BaseOperator.
+    """
+    def __call__(cls, *args, **kwargs):
+        """
+        Called when you call BaseOperator(). In this way we are able to perform an action
+        after initializing an operator no matter where  the ``super().__init__`` is called
+        (before or after assign of new attributes in a custom operator).
+        """
+        obj = type.__call__(cls, *args, **kwargs)
+        # Set upstream task defined by XComArgs passed to template fields of an operator
+        obj._set_xcomargs_dependencies()  # pylint: disable=protected-access
```

Comment:
meta_class introduce backward compatibility problems for users who already have metaclass defined for their operator ( they will have to inherit from BaseOperatorMeta) 
 
 what happens when user creates operator without DAG? 
 1. we can fix set_upstream or we can just finalize Operator implementation when we add it to dag ( thus no metaclass is required)

Reply:
> meta_class introduce backward compatibility problems for users who already have metaclass defined for their operator ( they will have to inherit from BaseOperatorMeta)

True

> what happens when a user creates an operator without DAG?

Well, nothing. If the question is about https://github.com/apache/airflow/pull/8805#discussion_r422633456 then AirflowException on missing DAG is raised. I've added a test for that.

> 1. we can fix set_upstream

Is this related to this comment https://github.com/apache/airflow/pull/8805#discussion_r422633321 ?

> 2. we can just finalize Operator implementation when we add it to dag ( thus no metaclass is required)

User can either provide the `dag` argument to operator or if it's `None` then we call `DagContext.get_current_dag()` . Both happens during `__init__` call, so I'm not sure if I understand when the "add to dag" moment happens.

Reply:
We can try to resolve those dependencies when creating a `DagRun`

Reply:
Additionally, I think we will have to introduce this metaclass now or later. Here's @mik-laj PR where he tried to address the `@apply_defaults`:
https://github.com/apache/airflow/pull/7450

One of the reasons to do that (from [devlist thread](https://lists.apache.org/thread.html/r4cf535e5dec1fbf8330f3c86031519092ca4820b382d3738413b23cd%40%3Cdev.airflow.apache.org%3E)):
> I saw that PR today and it made me realize I've not been using apply_defaults in much of my code! Look forward to this being an automaic operation via a MetaClass, great work :) 

However, this point is very valid:
> meta_class introduce backward compatibility problems for users who already have metaclass defined for their operator ( they will have to inherit from BaseOperatorMeta)

I suppose that we can add some refactor step once we have 2.0 and https://github.com/apache/airflow/issues/8765

Reply:
for this specific case (fixing apply_defaults probably should be done via metaclass) I would vote for dagbag "post" action. people can assign values not only after super().__init__ in subclass, but also in the DAG definition file as well. 
something like 
a = Op()
a.b = other.c

Reply:
> I would vote for dagbag "post" action. 

This would require some additional changes to make it work everywhere. For example, running `dag.run()` doesn't create DagBag thus the relation will not be set. 

> people can assign values not only after super().init in subclass, but also in the DAG definition file as well.

I am not sure I got it. To use `a.output` you have to first define `a` so there's where we resolve the upstream for the `a`. Using "functional" approach, users will have to define upstream tasks of an `a` task before he defines the `a` otherwise they will not be able to use `output` parameter.

Reply:
It is still possible to create tasks/ops without a dag, I.e. no argument and no "current" dag

Reply:
> > I would vote for dagbag "post" action. 
> 
> This would require some additional changes to make it work everywhere. For example, running `dag.run()` or `dag.create_dagrun()` don't create DagBag thus the relation will not be set. 
> 
> > people can assign values not only after super().init in subclass, but also in the DAG definition file as well.
> 
> Imho we should discourage that, I can't find any explanation why users should need to do that.
> 

This is python. Unless there is a _very_ good reason for not allowing it, we should let it work.

Reply:
> It is still possible to create tasks/ops without a dag, I.e. no argument and no "current" dag

Yes, but then you can not specify relations between tasks.

Reply:
I removed the metaclass